### PR TITLE
Ensure IPP objects are deallocated after failure and cancellation code paths with unit tests

### LINKED ISSUE #6612: [Mobile Payments] Weird payment/connection flow state
**Describe the bug**
When I attempt to collect a payment on a store with an empty store address and I switch to a different store IPP gets into a weird state.

**To Reproduce**
1. Switch to a site with an empty store address
2. Open detail of an unpaid order
3. Tap on collect payment
4. Notice "Missing store address" dialog is shown
5. Switch to a store with working IPP
6. Enter IPP settings
7. Connect to a reader
8. Disconnect from a reader 🔴 Notice the first tap on disconnect doesn't work -> try again
9. Try connect to a reader - 🔴  notice the connection keeps failing and `card_reader_connection_failed` - `Sorry, this payment couldn’t be processed` is tracked even though you are trying to connect to a reader not accept any payments.

There are more scenarios in which the app behaves unexpectedly - I think the key here is that when "missing store address" is shown, the app gets into a weird state - it doesn't cancel the payment flow perhaps? When I try switching stores and collecting a payment, the first attempt to collect the payment always fails.

**Mobile Environment**
Please include:
- Device: SIMULATOR
- iOS version: iPhone 13
- WooCommerce iOS version: dev `824da784749c47f4cac2659d6620b86e7ff73286`

### PR DESCRIPTION
**Pull request by jaclync**
Closes: #6612
Part of #5983 (unit tests for `CollectOrderPaymentUseCase` and `RefundSubmissionUseCase`)

### Description

First, an apology for going over 500 diffs in this PR 🙇🏻‍♀️ I could have separated the changes for the refund case to another PR, but thought it might be more clear to see the same changes to both payment and refund use cases.

#### The failure code path

There are several paths to reach the end of the IPP flow, encapsulated by `CollectOrderPaymentUseCase`. Since `CollectOrderPaymentUseCase` holds on to `CardReaderConnectionController` which subscribes to the reader state, we want to make sure all these IPP objects are deallocated after the end of the payment flow. Otherwise, any new updates on the card reader (like connecting to the reader after switching to a new store) could trigger the previous card reader subscription and result in weird behaviors like described in #6612. This is currently done at the completion block of `CollectOrderPaymentUseCase.collectPayment`:

https://github.com/woocommerce/woocommerce-ios/blob/667d162af29c682c8e383441b4bbeec0debece1d/WooCommerce/Classes/ViewModels/Order%20Details/OrderDetailsViewModel.swift#L542-L545

While adding unit tests for `CollectOrderPaymentUseCase` and `RefundSubmissionUseCase` in this PR, I noticed that the completion block for `CollectOrderPaymentUseCase.collectPayment` isn't invoked on the error case after dismissing an error alert. This is fixed by delegating the dismissal of the error alert from `CardPresentModalError` view model to the use case so that not only the alert UI is dismissed but also the completion block is called.

In addition to the completion block being invoked, there were retain cycles that caused the IPP objects to linger after the end of IPP flow. I took a pass to make sure to weak `self` in closures and the IPP objects are deallocated now.

#### The cancellation code path

Other than the failure scenario, I also noticed the completion block not being invoked after canceling the payment. This is now in a test case, and I also manually tested to make sure IPP objects are deallocated afterward using Xcode memory debugger.

#### Making payment/refund use case testable

In order to test the success/failure/cancellation scenarios of IPP, we need a way to interact with the alerts UI in unit tests. I added a protocol `OrderDetailsPaymentAlertsProtocol` so that we can pass the mock version to trigger different UI states. Please note that there are a lot more test cases for `CollectOrderPaymentUseCase` than `RefundSubmissionUseCase` due to the size of this PR. I will add more test cases when I implement analytics for the refund flow as part of https://github.com/woocommerce/woocommerce-ios/issues/5984.

### Testing instructions

Prerequisite: having two stores eligible for IPP
1. Switch to a site with an empty store address (you can update the store address in wp-admin > WooCommerce > Settings)
1. Go to the Orders tab and open an unpaid order
1. Tap on collect payment
1. Notice "Missing store address" alert is shown
1. Tap "Cancel" to dismiss the alert (before this PR, the completion block for this IPP flow isn't invoked due to how the error alert `CardPresentModalError` secondary action is set)
1. Switch to a store with working IPP
1. Enter IPP settings (Menu tab > Settings > In-Person Payments)
1. Connect to a reader --> the reader should be connected
1. Disconnect from a reader
1. Connect to a reader again --> the reader should be connected again
1. Go to the Orders tab and open an unpaid order
1. Collect payment --> the payment should go through

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by joshheald** [comment]
> While adding unit tests for CollectOrderPaymentUseCase and RefundSubmissionUseCase in this PR, I noticed that the completion block for CollectOrderPaymentUseCase.collectPayment isn't invoked on the error case after dismissing an error alert. This is fixed by delegating the dismissal of the error alert from CardPresentModalError view model to the use case so that not only the alert UI is dismissed but also the completion block is called.

Good catch. There's occasionally been strange behaviour that it sounds like this may fix... but I've never got repro steps for it before. Things like the payment prompt coming up when you connect a reader outside the payment flow, if you had difficulties connecting it at the time.

I'll keep an eye on whether they still happen, but this sounds really promising!

File: WooCommerce/Classes/ViewModels/CardPresentPayments/PaymentCaptureOrchestrator.swift [hunk 1]
```diff
@@ -309,21 +317,21 @@ private extension PaymentCaptureOrchestrator {
...
-                return String.localizedStringWithFormat(Localizations.belowMinimumAmount, amount)
+                return String.localizedStringWithFormat(Localization.belowMinimumAmount, amount)
             case .other:
-                return Localizations.defaultMessage
+                return Localization.defaultMessage
             }
         }
 
-        enum Localizations {
+        private enum Localization {
```
Comment (joshheald) [inline WooCommerce/Classes/ViewModels/CardPresentPayments/PaymentCaptureOrchestrator.swift:334]: Nice groundskeeping 🏕️

File: WooCommerce/Classes/ViewModels/Order Details/OrderDetailsPaymentAlerts.swift [hunk 2]
```diff
@@ -77,8 +77,8 @@ final class OrderDetailsPaymentAlerts {
         presentViewModel(viewModel: viewModel)
     }
 
-    func error(error: Error, tryAgain: @escaping () -> Void) {
-        let viewModel = errorViewModel(error: error, tryAgain: tryAgain)
+    func error(error: Error, tryAgain: @escaping () -> Void, dismissError: @escaping (UIViewController?) -> Void) {
```
Comment (joshheald) [inline WooCommerce/Classes/ViewModels/Order Details/OrderDetailsPaymentAlerts.swift]: What do you think about an alternative option here: make the passed in `dismissError` a dismiss completion handler instead. Then set the secondary action of the view model to a closure that calls `viewController.dismiss` with the `onCompletion` that was passed in.

That way you don't have to pass the view controller to the closure, nor handle the dismiss itself in several places.

File: WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift [hunk 3]
```diff
@@ -250,15 +265,15 @@ private extension RefundSubmissionUseCase {
                                              paymentGatewayAccount: paymentGatewayAccount,
                                              onWaitingForInput: { [weak self] in
             // Requests card input.
-            self?.alerts?.tapOrInsertCard(onCancel: {
+            self?.alerts.tapOrInsertCard(onCancel: { [weak self] in
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift:268]: Good spot 👍

File: WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift [hunk 4]
```diff
@@ -286,11 +301,14 @@ private extension RefundSubmissionUseCase {
...
-                    self.alerts?.nonRetryableError(from: self.rootViewController, error: cancelError)
+                    self.alerts.nonRetryableError(from: self.rootViewController, error: cancelError)
                     onCompletion(.failure(error))
                 }
             }
-        }
+        }, dismissError: { viewController in
+            viewController?.dismiss(animated: true)
+            onCompletion(.failure(error))
+        })
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift]: There will be a merge conflict with #6654 with this, but it should be easily resolved

**Review by joshheald — APPROVED** [review]
Tests well... I tried a few of the alerts I could think of with secondary actions, as well as your test instructions.

One suggestion for a tweak to the design – take it or leave it, and sorry if I missed something which meant that you need to do it this way!


So long, and thanks for all the tests!

File: WooCommerce/Classes/ViewModels/Order Details/OrderDetailsPaymentAlerts.swift (hunk 2 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewModels/Order Details/OrderDetailsPaymentAlerts.swift]: Great suggestion! Updated in https://github.com/woocommerce/woocommerce-ios/pull/6639/commits/9380bed2e961425d9ba815ef57b0dc92643fe127

File: WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift (hunk 3 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift:268]: It's so easy to miss one 😅

File: WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift (hunk 4 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift]: I can help with the merge conflicts after this PR is merged!

**Comment by jaclync** [comment]
Merging this PR now to make it easier for follow-up tasks!

**Pull request merged**